core/src/main/java/io/questdb/cutlass/pgwire/PGConnectionContext.java

```diff
@@ -1625,72 +1648,83 @@ private void parse(
-                // This parameterless message causes the backend to close the current transaction if it's not inside a BEGIN/COMMIT transaction block 
-                // (“close” meaning to commit if no error, or roll back if error). Then a ReadyForQuery response is issued. 
-                // The purpose of Sync is to provide a resynchronization point for error recovery. When an error is detected while processing any extended-query message, 
-                // the backend issues ErrorResponse, then reads and discards messages until a Sync is reached, then issues ReadyForQuery and returns to normal message processing. 
-                // (But note that no skipping occurs if an error is detected while processing Sync — this ensures that there is one and only one ReadyForQuery sent for each Sync.) 
-                processSyncActions();
-                prepareReadyForQuery();
-                prepareForNewQuery();
-                sendRNQ = true;
-                // fall thru
-            case 'H': // flush
-                // "The Flush message does not cause any specific output to be generated, but forces the backend to deliver any data pending in its output buffers. 
-                //  A Flush must be sent after any extended-query command except Sync, if the frontend wishes to examine the results of that command before issuing more commands. 
-                //  Without Flush, messages returned by the backend will be combined into the minimum possible number of packets to minimize network overhead."
-                // some clients (asyncpg) chose not to send 'S' (sync) message
-                // but instead fire 'H'. Can't wrap my head around as to why
-                // query execution is so ambiguous
-                if (syncActions.size() > 0) {
-                    processSyncActions();
-                }
-                sendAndReset();
-                break;
-            case 'D': // describe
-                sendRNQ = true;
-                processDescribe(msgLo, msgLimit, compiler);
-                break;
-            case 'Q': // simple query
-                sendRNQ = true;
-                processQuery(msgLo, msgLimit, compiler);
-                break;
-            case 'd': // COPY data 
-                break;
-            default:
-                LOG.error().$("unknown message [type=").$(type).$(']').$();
-                throw BadProtocolException.INSTANCE;
+
+        // after an error we only accept sync and flush messages. everything else is ignored. from the postgres protocol documentation:
+        // "[...] The purpose of Sync is to provide a resynchronization point for error recovery. When an error is detected
+        // while processing any extended-query message, the backend issues ErrorResponse, then reads and discards
+        // messages until a Sync is reached, then issues ReadyForQuery and returns to normal message processing. [...]"
+        if (!inErrorState || type == 'S' || type == 'H') {
```

Comment:
can we assume that all clients always send sync ? I don't think so, at least we need to validate this. If they don't, error never clears

Reply:
clients also disconnect and contexts are reused, this state needs also be cleared on disconnect. Otherwise clearing it on sync is probably ok, may be also on parse? When brand new query comes in - i don't really know. It feels very strict.

Reply:
> clients also disconnect and contexts are reused, this state needs also be cleared on disconnect. 

that's a good one! shouldn't we clear context state on each disconnect? error or not. 
we probably want to clear parsed queries, etc, don't we?

fwiw: if you feel uneasy about potentially disruptive nature of this changeset then the very first commit in this PR offers an alternative: https://github.com/questdb/questdb/pull/3237/commits/f70d561707d67d15b617c68e48893cb9054a9860 - this makes the rust sqlx client working. The rest of this PR is an attempt to make QuestDB more compliant with the protocol, but it's not without risks.

Reply:
right, if the fix is the one liner we should have that (that should be the first course of action next time). Rest is nice to have. To imlement protocol compliance we should source a test, client or anything else that helps us reproduce lack of compliance and create fix taget. Making changes to comply with a badly written documentation is a losing game.